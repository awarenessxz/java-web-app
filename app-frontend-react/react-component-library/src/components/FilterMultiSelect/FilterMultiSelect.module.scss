.text_secondary {
    color: rgba(0, 0, 0, 0.54);
    font-size: 0.875rem;
    font-family: "Roboto", "Helvetica", "Arial", sans-serif;
    font-weight: 400;
    line-height: 1.43;
    letter-spacing: 0.01071em;
}

.card {
    overflow: hidden;
    box-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2), 0px 1px 1px 0px rgba(0,0,0,0.14), 0px 1px 3px 0px rgba(0,0,0,0.12);
    border-radius: 4px;
    color: rgba(0, 0, 0, 0.87);
    transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
    background-color: #fff;

    & .card_header {
        padding: 8px 16px;
        display: flex;
        align-items: center;

        & .header_checkbox {
            flex: 0 0 auto;
            margin-right: 16px;
        }

        & .header_content {
            flex: 1 1 auto;

            & span {
                display:block;
            }
        }
    }

    & .divider {
        border: none;
        height: 1px;
        margin: 0;
        flex-shrink: 0;
        background-color: rgba(0, 0, 0, 0.12);
    }

    & .card_content {
        width: 100%;
        min-width: 200px;
        overflow: auto;
        background-color: #fff;
        margin: 0;
        position: relative;
        list-style: none;

        & .content_form_control {
            width: 100%;
            box-sizing: border-box;
            -webkit-box-sizing:border-box;
            -moz-box-sizing: border-box;
            text-transform: none;
            margin: 0;
            display: block;
            padding: .375rem .75rem;
            font-size: 1rem;
            line-height: 1.5;
            color: #495057;
            background-color: #fff;
            background-image: none;
            background-clip: padding-box;
            border: 1px solid #ced4da;
            border-radius: .25rem;
            -webkit-transition: border-color .15s ease-in-out,-webkit-box-shadow .15s ease-in-out;
            transition: border-color .15s ease-in-out,-webkit-box-shadow .15s ease-in-out;
            -o-transition: border-color ease-in-out .15s,box-shadow ease-in-out .15s;
            transition: border-color .15s ease-in-out,box-shadow .15s ease-in-out;
            transition: border-color .15s ease-in-out,box-shadow .15s ease-in-out,-webkit-box-shadow .15s ease-in-out;

            &:disabled {
                background: #d3d3d3;
            }

            &:focus {
                color: #495057;
                background-color: #fff;
                border-color: #80bdff;
                outline: 0;
                -webkit-box-shadow: 0 0 0 0.2rem rgba(0,123,255,.25);
                box-shadow: 0 0 0 0.2rem rgba(0,123,255,.25);
            }
        }

        & select {
            width: 100%;
        }
    }
}
