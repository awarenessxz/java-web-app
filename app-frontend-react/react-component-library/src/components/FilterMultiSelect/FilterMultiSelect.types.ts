export interface IHash {
    [key: string]: string;
}

export interface SelectOption {
    key: string;
    value: string;
}

export interface FilterMultiSelectProps {
    /** Options in list */
    options: SelectOption[];
    /** Selected options in list */
    selectedOptions: SelectOption[];
    /** callback function when selected options changes */
    onMultiSelectChange: (selected: SelectOption[]) => void;
    /** Title */
    title?: string;
    /** default filter for select */
    defaultFilter?: string;
    /** text in filter place holder */
    filterPlaceHolder?: string;
    /** enable filtering of select -- default = false */
    showFilter?: boolean;
    /** disable component */
    disabled?: boolean;
}
