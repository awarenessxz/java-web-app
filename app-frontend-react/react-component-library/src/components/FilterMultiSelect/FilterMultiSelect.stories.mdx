import { Meta, Story, ArgsTable } from '@storybook/addon-docs/blocks';
import FilterMultiSelect from './FilterMultiSelect';

<Meta title='Docs/FilterMultiSelect' component={FilterMultiSelect} />

# FilterMultiSelect

This component is built on top of html `select` and largely referenced from [react-filtered-multiselect](https://github.com/insin/react-filtered-multiselect).

<Story id="components-filtermultiselect--basic" />

# Usage

To use this component, do the following:

1. Add the library as dependency
    - `yarn add react-component-library`
2. Import the component
    ```javascript
    import React, { useState } from 'react';
    import { FilterMultiSelect, SelectOption } from 'react-component-library';

    ...
    const options = [
        { key: '1', value: 'text1' },
        { key: '2', value: 'text2' },
    ];
    const [selected, setSelected] = useState<SelectOption[]>([]);

    const handleMultiSelectChange = (selection: SelectOption[]): void => {
        console.log(selection);
        setSelected(selection);
    };

    return (
        <FilterMultiSelect
            options
            selectedOptions: selected
            onMultiSelectChange: handleMultiSelectChange
        />
    );
    ...
    ```

# Properties

<ArgsTable of={FilterMultiSelect} />

# Typescript Types

```
interface SelectOption {
    key: string;
    value: string;
}
```


