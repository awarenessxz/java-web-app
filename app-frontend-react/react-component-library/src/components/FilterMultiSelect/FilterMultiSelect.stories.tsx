import * as React from 'react';
import { Meta, Story } from '@storybook/react/types-6-0';
import FilterMultiSelect from './FilterMultiSelect';
import { FilterMultiSelectProps } from './FilterMultiSelect.types';

export default {
    title: 'Components/FilterMultiSelect',
    component: FilterMultiSelect,
    excludeStories: /.*Data$/,
} as Meta;

// create a template
const Template: Story<FilterMultiSelectProps> = (args) => {
    return <FilterMultiSelect {...args} />;
};

/** *************************************************
 * Data
 ************************************************** */

const sampleArgs: FilterMultiSelectProps = {
    options: [
        {
            key: 'value1',
            value: 'Value 1',
        },
        {
            key: 'value2',
            value: 'Value 2',
        },
        {
            key: 'value3',
            value: 'Value 3',
        },
        {
            key: 'value4',
            value: 'Value 4',
        },
        {
            key: 'value5',
            value: 'Value 5',
        },
    ],
    selectedOptions: [
        {
            key: 'value4',
            value: 'Value 4',
        },
    ],
    onSelected: () => {
        console.log('Selected!!');
    },
    showFilter: true,
};

/** *************************************************
 * Stories
 ************************************************** */

// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
export const Basic: Story = Template.bind({});
Basic.args = { ...sampleArgs };
