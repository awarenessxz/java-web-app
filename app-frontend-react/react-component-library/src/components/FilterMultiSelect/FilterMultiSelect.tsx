import React, { ChangeEvent, useRef, useState } from 'react';
import { FilterMultiSelectProps } from './FilterMultiSelect.types';
import { filterOptions, isSelectedOptionsUpdated } from './FilterMultiSelect.util';
import styles from './FilterMultiSelect.module.scss';

/**
 * FilterMultiSelect
 */
const FilterMultiSelect = ({
    options = [],
    selectedOptions = [],
    onSelectChange,
    title = undefined,
    defaultFilter = '',
    filterPlaceHolder = 'type to filter',
    showFilter = false,
    disabled = false,
}: FilterMultiSelectProps): JSX.Element => {
    const [mFilter, setmFilter] = useState(defaultFilter);
    const [mFilteredOptions, setmFilteredOptions] = useState(filterOptions(defaultFilter, options));
    const [mSelectedOptions, setmSelectedOptions] = useState(selectedOptions);
    const checkboxAllRef = useRef<HTMLInputElement>(null);
    const selectRef = useRef<HTMLSelectElement>(null);

    const updateSelectedValue = (e?: ChangeEvent<HTMLSelectElement>): void => {
        const el = e ? e.target : selectRef.current;
        if (el) {
            const newSelectedOptions = [];
            for (let i = 0, l = el.options.length; i < l; i += 1) {
                if (el.options[i].selected) {
                    newSelectedOptions.push({
                        key: el.options[i].value,
                        value: el.options[i].text,
                    });
                }
            }
            if (checkboxAllRef && checkboxAllRef.current) {
                checkboxAllRef.current.checked = newSelectedOptions.length === el.options.length;
            }

            // Always update if we were handling an event, otherwise only update if selectedValues has actually changed.
            if (e || isSelectedOptionsUpdated(mSelectedOptions, newSelectedOptions)) {
                setmSelectedOptions(newSelectedOptions);
                onSelectChange(newSelectedOptions);
            }
        }
    };

    const onCheckAllToggle = (e: ChangeEvent<HTMLInputElement>): void => {
        const { checked } = e.target;
        if (checked) {
            if (selectRef && selectRef.current) {
                for (let i = 0, l = selectRef.current.options.length; i < l; i += 1) {
                    selectRef.current.options[i].selected = checked;
                }
            }
        } else {
            setmSelectedOptions([]);
        }
    };

    const onFilterChange = (e: ChangeEvent<HTMLInputElement>): void => {
        const filter = e.target.value;
        setmFilter(filter);
        setmFilteredOptions(filterOptions(filter, options));
    };

    return (
        <div className={styles.card}>
            <div className={styles.card_header}>
                <div className={styles.header_checkbox}>
                    <input ref={checkboxAllRef} type="checkbox" onChange={onCheckAllToggle} disabled={disabled} />
                </div>
                <div className={styles.header_content}>
                    <span>{title}</span>
                    <span className={styles.text_secondary}>
                        {mSelectedOptions.length}/{options.length} selected
                    </span>
                </div>
            </div>
            <hr className={styles.divider} />
            <div className={styles.card_content}>
                {showFilter && (
                    <input
                        type="text"
                        placeholder={filterPlaceHolder}
                        className={styles.content_form_control}
                        onChange={onFilterChange}
                        value={mFilter}
                        disabled={disabled}
                    />
                )}
                <select
                    multiple
                    ref={selectRef}
                    className={styles.content_form_control}
                    onChange={updateSelectedValue}
                    size={mFilteredOptions.length}
                    value={mSelectedOptions.map((option) => option.key)}
                    disabled={disabled}
                >
                    {mFilteredOptions.map((option) => {
                        return (
                            <option key={option.key} value={option.key}>
                                {option.value}
                            </option>
                        );
                    })}
                </select>
            </div>
        </div>
    );
};

export default FilterMultiSelect;
