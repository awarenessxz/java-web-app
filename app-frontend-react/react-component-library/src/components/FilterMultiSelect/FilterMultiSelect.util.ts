import { IHash, SelectOption } from './FilterMultiSelect.types';

// filter options by filter text
export const filterOptions = (filter: string, options: SelectOption[]): SelectOption[] => {
    const mFilter = filter.toUpperCase();
    const mFilteredOptions: SelectOption[] = [];
    for (let i = 0, l = options.length; i < l; i += 1) {
        if (!mFilter || options[i].value.toUpperCase().indexOf(mFilter) !== -1) {
            mFilteredOptions.push(options[i]);
        }
    }
    return mFilteredOptions;
};

interface ProcessSelectedOptionsResult {
    visible: SelectOption[];
    hidden: SelectOption[];
}

// split selected options into hidden selected options and visible selected options
export const processSelectedOptions = (
    selectedOptions: SelectOption[],
    filteredOptions: SelectOption[],
): ProcessSelectedOptionsResult => {
    const visibleSelectedOptions: SelectOption[] = [];
    const hiddenSelectedOptions: SelectOption[] = [];
    selectedOptions.forEach((option) => {
        let found = false;
        for (let i = 0; i < filteredOptions.length; i += 1) {
            if (filteredOptions[i].key === option.key) {
                found = true;
                break;
            }
        }
        if (found) {
            visibleSelectedOptions.push(option);
        } else {
            hiddenSelectedOptions.push(option);
        }
    });
    return { visible: visibleSelectedOptions, hidden: hiddenSelectedOptions };
};

// get hash map of selected Options
export const convertSelectedOptionsToHashmap = (options: SelectOption[]): IHash =>
    Object.assign({}, ...options.map((option) => ({ [option.key]: option.value }))) as IHash;

// check if selected options is updated
export const isSelectedOptionsUpdated = (original: SelectOption[], newOptions: SelectOption[]): boolean => {
    const keysA = original.map((option) => option.key);
    const keysB = newOptions.map((option) => option.key);
    return String(keysA) !== String(keysB);
};
