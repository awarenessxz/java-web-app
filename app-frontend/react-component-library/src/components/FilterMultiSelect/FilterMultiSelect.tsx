import React, { ChangeEvent, useEffect, useRef, useState } from 'react';
import { FilterMultiSelectProps, SelectOption, SelectedOptions } from './FilterMultiSelect.types';
import { filterOptions, processSelectedOptions } from './FilterMultiSelect.util';
import styles from './FilterMultiSelect.module.scss';

/**
 * FilterMultiSelect
 */
const FilterMultiSelect = ({
    options = [],
    selectedOptions = [],
    onMultiSelectChange,
    title = undefined,
    defaultFilter = '',
    filterPlaceHolder = 'type to filter',
    showFilter = true,
    disabled = false,
    ...props
}: FilterMultiSelectProps): JSX.Element => {
    const [mFilter, setmFilter] = useState(defaultFilter);
    const [mFilteredOptions, setmFilteredOptions] = useState<SelectOption[]>([]); // keeps track of current visible options
    const [mSelectedOptions, setmSelectedOptions] = useState<SelectedOptions>({
        visible: [], // keeps track of current visible selected options
        hidden: [], // keeps track of selected options that are hidden away due to filter
    });
    const checkboxAllRef = useRef<HTMLInputElement>(null);
    const selectRef = useRef<HTMLSelectElement>(null);
    const triggerMultiChange = useRef(false);

    // init on first load
    useEffect(() => {
        const filteredOptions = filterOptions(defaultFilter, options);
        const { visible, hidden } = processSelectedOptions(selectedOptions, filteredOptions);
        setmFilteredOptions(filteredOptions);
        setmSelectedOptions({ visible, hidden });
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    // update checkbox ui check state
    const updateCheckboxState = (): void => {
        if (checkboxAllRef && checkboxAllRef.current) {
            const selectedLength = mSelectedOptions.visible.length + mSelectedOptions.hidden.length;
            checkboxAllRef.current.checked = selectedLength === options.length;
        }
    };

    // trigger this whenever selected options is updated
    useEffect(() => {
        // only useEffect when the component have been init
        if (triggerMultiChange.current) {
            // callback when selection change
            const totalSelectedOptions = [...mSelectedOptions.visible, ...mSelectedOptions.hidden];
            onMultiSelectChange(totalSelectedOptions);
            triggerMultiChange.current = false;
        }
        updateCheckboxState();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [mSelectedOptions]);

    // handle selection changes
    const onSelectChange = (e: ChangeEvent<HTMLSelectElement>): void => {
        const newSelectedOptions: SelectOption[] = [];
        Array.from(e.target.selectedOptions, (option) => {
            newSelectedOptions.push({
                key: option.value,
                value: option.text,
            });
            return option.value; // not used
        });

        // update state (trigger effect)
        triggerMultiChange.current = true;
        setmSelectedOptions({ ...mSelectedOptions, visible: newSelectedOptions });
    };

    // handle checkbox changes
    const onCheckAllToggle = (e: ChangeEvent<HTMLInputElement>): void => {
        const { checked } = e.target;
        if (checked) {
            if (selectRef && selectRef.current) {
                // add those visible non-selected options to selected
                const newSelectedOptions: SelectOption[] = [];
                for (let i = 0, l = selectRef.current.options.length; i < l; i += 1) {
                    if (!selectRef.current.options[i].selected) {
                        newSelectedOptions.push({
                            key: selectRef.current.options[i].value,
                            value: selectRef.current.options[i].text,
                        });
                    }
                }
                // update state (trigger effect)
                triggerMultiChange.current = true;
                setmSelectedOptions({
                    ...mSelectedOptions,
                    visible: [...mSelectedOptions.visible, ...newSelectedOptions],
                });
            }
        } else {
            // remove all selected options, update state (trigger effect)
            triggerMultiChange.current = true;
            setmSelectedOptions({ ...mSelectedOptions, visible: [] });
        }
    };

    const onFilterChange = (e: ChangeEvent<HTMLInputElement>): void => {
        const filter = e.target.value;
        const filteredOptions = filterOptions(filter, options);
        const allSelectedOptions = [...mSelectedOptions.visible, ...mSelectedOptions.hidden];
        const { visible, hidden } = processSelectedOptions(allSelectedOptions, filteredOptions);
        setmFilter(filter);
        setmFilteredOptions(filteredOptions);
        setmSelectedOptions({ visible, hidden });
    };

    return (
        <div className={styles.card} {...props}>
            <div className={styles.card_header}>
                <div className={styles.header_checkbox}>
                    <input
                        ref={checkboxAllRef}
                        data-testid="input-checkbox"
                        type="checkbox"
                        onChange={onCheckAllToggle}
                        disabled={disabled}
                    />
                </div>
                <div className={styles.header_content}>
                    <span>{title}</span>
                    <span className={styles.text_secondary}>
                        {mSelectedOptions.visible.length + mSelectedOptions.hidden.length}/{options.length} selected
                    </span>
                </div>
            </div>
            <hr className={styles.divider} />
            <div className={styles.card_content}>
                {showFilter && (
                    <input
                        type="text"
                        data-testid="input-filter"
                        placeholder={filterPlaceHolder}
                        className={styles.content_form_control}
                        onChange={onFilterChange}
                        value={mFilter}
                        disabled={disabled}
                    />
                )}
                <select
                    multiple
                    ref={selectRef}
                    data-testid="select-multiple"
                    className={styles.content_form_control}
                    onChange={onSelectChange}
                    size={mFilteredOptions.length}
                    value={mSelectedOptions.visible.map((option) => option.key)}
                    disabled={disabled}
                >
                    {mFilteredOptions.map((option) => {
                        return (
                            <option data-testid="select-multiple-option" key={option.key} value={option.key}>
                                {option.value}
                            </option>
                        );
                    })}
                </select>
            </div>
        </div>
    );
};

export default FilterMultiSelect;
