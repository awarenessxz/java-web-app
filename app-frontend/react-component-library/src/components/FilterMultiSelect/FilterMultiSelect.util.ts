import { SelectOption, SelectedOptions } from './FilterMultiSelect.types';

// filter options by filter text
export const filterOptions = (filter: string, options: SelectOption[]): SelectOption[] => {
    const mFilter = filter.toUpperCase();
    const mFilteredOptions: SelectOption[] = [];
    for (let i = 0, l = options.length; i < l; i += 1) {
        if (!mFilter || options[i].value.toUpperCase().indexOf(mFilter) !== -1) {
            mFilteredOptions.push(options[i]);
        }
    }
    return mFilteredOptions;
};

// split selected options into hidden selected options and visible selected options
export const processSelectedOptions = (
    selectedOptions: SelectOption[],
    filteredOptions: SelectOption[],
): SelectedOptions => {
    const visibleSelectedOptions: SelectOption[] = [];
    const hiddenSelectedOptions: SelectOption[] = [];
    selectedOptions.forEach((option) => {
        let found = false;
        for (let i = 0; i < filteredOptions.length; i += 1) {
            if (filteredOptions[i].key === option.key) {
                found = true;
                break;
            }
        }
        if (found) {
            visibleSelectedOptions.push(option);
        } else {
            hiddenSelectedOptions.push(option);
        }
    });
    return { visible: visibleSelectedOptions, hidden: hiddenSelectedOptions };
};
