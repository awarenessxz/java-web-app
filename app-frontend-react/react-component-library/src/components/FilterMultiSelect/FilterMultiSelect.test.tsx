import * as React from 'react';
import { render, RenderResult } from '@testing-library/react';

// Import Component
import FilterMultiSelect from './FilterMultiSelect';
import { FilterMultiSelectProps } from './FilterMultiSelect.types';

// function to render Component before each test
const renderComponent = (props = {}): RenderResult => {
    const defaultProps: FilterMultiSelectProps = {
        title: 'FilterMultiSelect',
    };
    const merged = { ...defaultProps, ...props };
    return render(<FilterMultiSelect {...merged} />);
};

// 1. Testing if component renders properly
describe('Testing if component renders properly', () => {
    // Snapshot Testing
    it('Snapshot Testing', () => {
        const { asFragment } = renderComponent();
        expect(asFragment()).toMatchSnapshot();
    });
});

