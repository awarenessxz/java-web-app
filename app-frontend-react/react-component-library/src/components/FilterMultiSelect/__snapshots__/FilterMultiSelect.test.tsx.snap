// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing if component renders properly Snapshot Testing 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="card_header"
    >
      <div
        class="header_checkbox"
      >
        <input
          data-testid="input-checkbox"
          type="checkbox"
        />
      </div>
      <div
        class="header_content"
      >
        <span />
        <span
          class="text_secondary"
        >
          0/2 selected
        </span>
      </div>
    </div>
    <hr
      class="divider"
    />
    <div
      class="card_content"
    >
      <input
        class="content_form_control"
        data-testid="input-filter"
        placeholder="type to filter"
        type="text"
        value=""
      />
      <select
        class="content_form_control"
        data-testid="select-multiple"
        multiple=""
        size="2"
      >
        <option
          data-testid="select-multiple-option"
          value="value1"
        >
          Value 1
        </option>
        <option
          data-testid="select-multiple-option"
          value="value2"
        >
          Value 2
        </option>
      </select>
    </div>
  </div>
</DocumentFragment>
`;
